File: src/main/java/com/fwdekker/randomness/uuid/UuidSettingsComponent.java
package com.fwdekker.randomness.uuid;

import com.fwdekker.randomness.SettingsComponent;
import com.fwdekker.randomness.ui.ButtonGroupKt;
import com.intellij.openapi.ui.ValidationInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.ButtonGroup;
import javax.swing.JPanel;


/**
 * Component for settings of random UUID generation.
 *
 * @see UuidSettings
 * @see UuidSettingsAction
 */
public final class UuidSettingsComponent extends SettingsComponent<UuidSettings> {
    private JPanel contentPane;
    private ButtonGroup enclosureGroup;


    /**
     * Constructs a new {@code UuidSettingsComponent} that uses the singleton {@code UuidSettings} instance.
     */
    /* default */ UuidSettingsComponent() {
        this(UuidSettings.Companion.getDefault());
    }

    /**
     * Constructs a new {@code UuidSettingsComponent} that uses the given {@code UuidSettings} instance.
     *
     * @param settings the settings to manipulate with this component
     */
    /* default */ UuidSettingsComponent(final @NotNull UuidSettings settings) {
        super(settings);

        loadSettings();
    }


    @Override
    public JPanel getRootPane() {
        return contentPane;
    }


    @Override
    public void loadSettings(final @NotNull UuidSettings settings) {
        ButtonGroupKt.setValue(enclosureGroup, settings.getEnclosure());
    }

    @Override
    public void saveSettings(final @NotNull UuidSettings settings) {
        final String enclosure = ButtonGroupKt.getValue(enclosureGroup);
        settings.setEnclosure(enclosure == null ? UuidSettings.DEFAULT_ENCLOSURE : enclosure);
    }

    @Override
    @Nullable
    public ValidationInfo doValidate() {
        return null;
    }
}

File: src/main/kotlin/com/fwdekker/randomness/uuid/UuidSettingsComponent.kt
package com.fwdekker.randomness.uuid

import com.fwdekker.randomness.CapitalizationMode.Companion.getMode
import com.fwdekker.randomness.SchemeComponent
import com.fwdekker.randomness.ValidationInfo
import com.fwdekker.randomness.ui.addChangeListenerTo
import com.fwdekker.randomness.ui.getValue
import com.fwdekker.randomness.ui.setValue
import com.fwdekker.randomness.uuid.UuidScheme.Companion.DEFAULT_CAPITALIZATION
import com.fwdekker.randomness.uuid.UuidScheme.Companion.DEFAULT_ENCLOSURE
import com.fwdekker.randomness.uuid.UuidScheme.Companion.DEFAULT_VERSION
import javax.swing.ButtonGroup
import javax.swing.JCheckBox
import javax.swing.JPanel


/**
 * Component for settings of random UUID generation.
 *
 * @param settings the settings to edit in the component
 */
@Suppress("LateinitUsage") // Initialized by scene builder
class UuidSettingsComponent(settings: UuidScheme) : SchemeComponent<UuidScheme>(settings) {
    private lateinit var contentPane: JPanel
    private lateinit var versionGroup: ButtonGroup
    private lateinit var enclosureGroup: ButtonGroup
    private lateinit var capitalizationGroup: ButtonGroup
    private lateinit var addDashesCheckBox: JCheckBox

    override val rootPane get() = contentPane


    init {
        loadScheme()
    }


    override fun loadScheme(scheme: UuidScheme) {
        versionGroup.setValue(scheme.version.toString())
        enclosureGroup.setValue(scheme.enclosure)
        capitalizationGroup.setValue(scheme.capitalization)
        addDashesCheckBox.isSelected = scheme.addDashes
    }

    override fun saveScheme(scheme: UuidScheme) {
        scheme.version = versionGroup.getValue()?.toInt() ?: DEFAULT_VERSION
        scheme.enclosure = enclosureGroup.getValue() ?: DEFAULT_ENCLOSURE
        scheme.capitalization = capitalizationGroup.getValue()?.let { getMode(it) } ?: DEFAULT_CAPITALIZATION
        scheme.addDashes = addDashesCheckBox.isSelected
    }

    override fun doValidate(): ValidationInfo? = null

    override fun addChangeListener(listener: () -> Unit) =
        addChangeListenerTo(
            versionGroup, enclosureGroup, capitalizationGroup, addDashesCheckBox,
            listener = listener
        )

    override fun toUDSDescriptor() = "%UUID[]"
}
